projects/weight_init/weight_init_experiment.py

```diff
@@ -0,0 +1,671 @@
+            ret["noise_accuracy"] = ret["noise_accuracies"][-1]
+            ret["mean_loss"] = loss / len(noise_values)
+
+            self.last_noise_results = ret
+
+        return ret
+
+    def _post_epoch(self, epoch):
+        """
+        The set of actions to do after each epoch of training: adjust learning
+        rate, rezero sparse weights, and update boost strengths.
+        """
+        self._adjust_learning_rate(self.optimizer, epoch)
+        self.model.apply(rezero_weights)
+        self.model.apply(update_boost_strength)
+
+    def _initialize_weights(self):
+        if self.weight_init == "lsuv":
+            initializer = LSUVWeightInit(
+                self.model, self.train_loader, device=self.device
+            )
+            initializer.initialize()
+        elif self.weight_init == "grassmannian":
+            initializer = GrassmannianWeightInit(self.model, device=self.device)
+            initializer.initialize()
+        elif self.weight_init == "default":
+            for m in self.model.modules():
+                if isinstance(m, nn.Conv2d):
+                    n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
+                    m.weight.data.normal_(0, math.sqrt(2.0 / n))
+                    if m.bias is not None:
+                        m.bias.data.zero_()
+                elif isinstance(m, nn.BatchNorm2d):
+                    m.weight.data.fill_(1)
+                    m.bias.data.zero_()
+                elif isinstance(m, nn.Linear):
+                    n = m.weight.size(1)
+                    m.weight.data.normal_(0, 0.01)
+                    m.bias.data.zero_()
+
+
+class GrassmannianWeightInit(object):
+    """
+    Set first conv layer weights to Grassmannian
+    From the paper by J. H. Conway, R. H. Hardin and N. J. A. Sloane
+    """
+
+    def __init__(self, model, device=None):
+        self.model = model
+        self.device = device
+
+    def grassmannian_extract(self, n, k_size, num_channels):
+        """
+        Download packed subspace from Sloane (http://neilsloane.com/grass/)
+        """
+        target_url = "http://neilsloane.com/grass/dim{}/grassc.{}.{}.{}.txt".format(
+            k_size, k_size, num_channels, n
+        )
+        response = requests.get(target_url)
+        list_str = response.text.split("\n")[0 : num_channels * k_size * n]
+        k = int(np.sqrt(k_size))
+        w_mat = np.float_(list_str).reshape(n, k, k, num_channels)
+        print("Downloaded weights of shape: %s" % str(w_mat.shape))
+        return w_mat
+
+    def initialize(self):
+        for m in self.model.modules():
+            if isinstance(m, nn.Conv2d):
+                n_filters = m.out_channels
+                n_input = m.in_channels
+                k = m.kernel_size[0]
+                w = self.grassmannian_extract(n_filters, k * k, n_input)
+                m.weight.data = torch.from_numpy(w).float().to(self.device)
+                # Break after first Conv layer
```

Comment:
Why only the first layer?

Reply:
Prabhu's paper introducing Grassmannian used pre-packed weights only for the first layer: 
"With the intuition that the first few convolutional kernelfilters should capture as much diverse features as possibleto prevent ‘dead kernels’ (Zeiler & Fergus, 2014)..."

May be interesting to try with all layers as well.